README
NAME
    POE::Component::Client::Whois - A one shot non-blocking RFC 812 WHOIS
    query.

VERSION
    version 1.30

SYNOPSIS
       use strict;
       use warnings;
       use POE qw(Component::Client::Whois);
       use Data::Dumper;

       POE::Session->create(
            package_states => [
                    'main' => [ qw(_start _response) ],
            ],
       );

       $poe_kernel->run();
       exit 0;

       sub _start {
         my ($kernel,$heap) = @_[KERNEL,HEAP];

         POE::Component::Client::Whois->whois( host => "whois.nic.uk",
                                               query => 'bingosnet.co.uk',
                                               event => '_response',
                                               _arbitary => [ qw(moo moo moo) ] );
         undef;
       }

       sub _response {
          print STDERR Dumper( $_[ARG0] );
       }

DESCRIPTION
    POE::Component::Client::Whois provides a lightweight one shot
    non-blocking RFC 812 WHOIS query to other POE sessions and components.
    The component will attempt to guess the appropriate whois server to
    connect to based on the query string passed.

    If no guess can be made it will connect to whois.internic.net for
    domains, whois.arin.net for IPv4 addresses and whois.6bone.net for IPv6
    addresses.

CONSTRUCTOR
    "whois"
        Creates a POE::Component::Client::Whois session. Takes two mandatory
        arguments and a number of optional:

          'query', the string query to send to the whois server; # Mandatory
          'event', the event name to emit on success/failure; # Mandatory
          'port', the port on the whois server to connect to, defaults to 43;
          'session', a session or alias to send the above 'event' to, defaults to calling session;
          'host', the whois server to query; # Automagically determined by the component
          'referral', indicates to the poco whether to follow ReferralServer; # Default is 1
          'referral_only', indicates whether the poco should only return the results of querying
                           the referral server; # default is 0

        One can also pass arbitary data to whois() which will be passed back
        in the response event. It is advised that one uses an underscore
        prefix to avoid clashes with future versions.

OUTPUT
    ARG0 will be a hashref, which contains the original parameters passed to
    whois() ( including any arbitary data ), plus either one of the
    following two keys:

      'reply', an arrayref of response lines from the whois server, assuming no error occurred;
      'error', in lieu of a valid response, this will be defined with a brief description of
               what went wrong;

    No parsing is undertaken on the returned data, this is an exercise left
    to the reader >;]

KUDOS
    ketas, for first suggesting this module; buu, decay and hazard from
    #perl @ freenode, for helpful suggestions;

    This module is based on the linux whois client from
    <http://www.linux.it/~md/software/>.

SEE ALSO
    RFC 812 <http://www.faqs.org/rfcs/rfc812.html>.

AUTHOR
    Chris Williams <chris@bingosnet.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Chris Williams.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2008-10-20 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 1.30 at 2013-10-19 18:28:10 +0000
-----------------------------------------

  Change: 74c6fb8c2402d3f08a65630245450542f5973853
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-10-19 19:28:10 +0000

    Change distribution to Dist::Zilla 

  Change: 2cd6ae2971a0d6722d97b5c438aba9d0bb6ecbc5
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-10-19 19:17:30 +0000

    Updated ipblks 

  Change: 2812b9185f25691cf7f20af51c5ca5cc4480e914
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-08-28 10:31:42 +0000

    Update IPBlks data 

-----------------------------------------
version 1.28 at 2009-10-31 12:26:57 +0000
-----------------------------------------

  Change: dc5267887b9ef9a69eaa92176656f7665241787f
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-10-31 12:26:57 +0000

    Actually fixed referral following 

-----------------------------------------
version 1.26 at 2009-10-31 11:24:08 +0000
-----------------------------------------

  Change: 1155905bd3061444c4944702b4bc55a11daa8a58
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-10-31 11:24:08 +0000

    Fixed referral following 

-----------------------------------------
version 1.24 at 2009-10-15 19:47:52 +0000
-----------------------------------------

  Change: 04ff7c738271c94161ce2e83bfe37b9f31bdbfcf
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-10-15 20:47:52 +0000

    Implemented NONE and WEB handling behaviour 

-----------------------------------------
version 1.22 at 2009-10-15 13:21:26 +0000
-----------------------------------------

  Change: f7d906ce59eba9dd134a9bd5f9e7c9993aebe093
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-10-15 14:21:26 +0000

    Added another test to TLD for WEB 

  Change: 4d7e84c93420f0aa72fdf257e4607d3564fcd997
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-10-15 14:18:28 +0000

    Refactoring of the TLD lookups. Updated TLD lists 

-----------------------------------------
version 1.20 at 2009-04-08 20:26:28 +0000
-----------------------------------------

  Change: 075840bcaa72139e06c008c28a047256adb4977a
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-08 21:26:28 +0000

    Auto set repository in META.yml. Adjust some tests to try and track
    down fails 

  Change: 66cd34e0eabd834a7f94dd66c05e2ce3a86c3c04
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-01 13:17:07 +0000

    Initial git import 

================
End of releases.
================


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A one shot non-blocking RFC 812 WHOIS query.",
  "AUTHOR" => "Chris Williams <chris\@bingosnet.co.uk>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "POE-Component-Client-Whois",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "POE::Component::Client::Whois",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Net::IP::Minimal" => 0,
    "Net::Netmask" => "1.9012",
    "POE" => "1.004",
    "POE::Filter::Line" => 0,
    "POE::Wheel::ReadWrite" => 0,
    "POE::Wheel::SocketFactory" => 0,
    "Socket" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.30",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/POE/Component/Client/Whois/IPBlks.pm
package POE::Component::Client::Whois::IPBlks;
{
  $POE::Component::Client::Whois::IPBlks::VERSION = '1.30';
}

#ABSTRACT: Determine which whois server is responsible for a network address.

use strict;
use warnings;
use Net::Netmask;

sub new {
  my $self = bless { }, shift;
  $self->{data} = {
  '1.0.0.0/8' => 'whois.apnic.net',
  '2.0.0.0/8' => 'whois.ripe.net',
  '3.0.0.0/8' => 'whois.arin.net',
  '4.0.0.0/8' => 'whois.arin.net',
  '5.0.0.0/8' => 'whois.ripe.net',
  '7.0.0.0/8' => 'whois.arin.net',
  '8.0.0.0/8' => 'whois.arin.net',
  '9.0.0.0/8' => 'whois.arin.net',
  '12.0.0.0/8' => 'whois.arin.net',
  '13.0.0.0/8' => 'whois.arin.net',
  '14.0.0.0/8' => 'whois.apnic.net',
  '15.0.0.0/8' => 'whois.arin.net',
  '16.0.0.0/8' => 'whois.arin.net',
  '17.0.0.0/8' => 'whois.arin.net',
  '18.0.0.0/8' => 'whois.arin.net',
  '19.0.0.0/8' => 'whois.arin.net',
  '20.0.0.0/8' => 'whois.arin.net',
  '23.0.0.0/8' => 'whois.arin.net',
  '24.0.0.0/8' => 'whois.arin.net',
  '25.0.0.0/8' => 'whois.ripe.net',
  '27.0.0.0/8' => 'whois.apnic.net',
  '31.0.0.0/8' => 'whois.ripe.net',
  '32.0.0.0/8' => 'whois.arin.net',
  '34.0.0.0/8' => 'whois.arin.net',
  '35.0.0.0/8' => 'whois.arin.net',
  '36.0.0.0/8' => 'whois.apnic.net',
  '37.0.0.0/8' => 'whois.ripe.net',
  '38.0.0.0/8' => 'whois.arin.net',
  '39.0.0.0/8' => 'whois.apnic.net',
  '40.0.0.0/8' => 'whois.arin.net',
  '41.0.0.0/8' => 'whois.afrinic.net',
  '42.0.0.0/8' => 'whois.apnic.net',
  '43.0.0.0/8' => 'whois.apnic.net',
  '44.0.0.0/8' => 'whois.arin.net',
  '45.0.0.0/8' => 'whois.arin.net',
  '46.0.0.0/8' => 'whois.ripe.net',
  '47.0.0.0/8' => 'whois.arin.net',
  '48.0.0.0/8' => 'whois.arin.net',
  '49.0.0.0/8' => 'whois.apnic.net',
  '50.0.0.0/8' => 'whois.arin.net',
  '51.0.0.0/8' => 'whois.ripe.net',
  '52.0.0.0/8' => 'whois.arin.net',
  '54.0.0.0/8' => 'whois.arin.net',
  '56.0.0.0/8' => 'whois.arin.net',
  '58.0.0.0/8' => 'whois.apnic.net',
  '59.0.0.0/8' => 'whois.apnic.net',
  '60.0.0.0/8' => 'whois.apnic.net',
  '61.0.0.0/8' => 'whois.apnic.net',
  '62.0.0.0/8' => 'whois.ripe.net',
  '63.0.0.0/8' => 'whois.arin.net',
  '64.0.0.0/8' => 'whois.arin.net',
  '65.0.0.0/8' => 'whois.arin.net',
  '66.0.0.0/8' => 'whois.arin.net',
  '67.0.0.0/8' => 'whois.arin.net',
  '68.0.0.0/8' => 'whois.arin.net',
  '69.0.0.0/8' => 'whois.arin.net',
  '70.0.0.0/8' => 'whois.arin.net',
  '71.0.0.0/8' => 'whois.arin.net',
  '72.0.0.0/8' => 'whois.arin.net',
  '73.0.0.0/8' => 'whois.arin.net',
  '74.0.0.0/8' => 'whois.arin.net',
  '75.0.0.0/8' => 'whois.arin.net',
  '76.0.0.0/8' => 'whois.arin.net',
  '77.0.0.0/8' => 'whois.ripe.net',
  '78.0.0.0/8' => 'whois.ripe.net',
  '79.0.0.0/8' => 'whois.ripe.net',
  '80.0.0.0/8' => 'whois.ripe.net',
  '81.0.0.0/8' => 'whois.ripe.net',
  '82.0.0.0/8' => 'whois.ripe.net',
  '83.0.0.0/8' => 'whois.ripe.net',
  '84.0.0.0/8' => 'whois.ripe.net',
  '85.0.0.0/8' => 'whois.ripe.net',
  '86.0.0.0/8' => 'whois.ripe.net',
  '87.0.0.0/8' => 'whois.ripe.net',
  '88.0.0.0/8' => 'whois.ripe.net',
  '89.0.0.0/8' => 'whois.ripe.net',
  '90.0.0.0/8' => 'whois.ripe.net',
  '91.0.0.0/8' => 'whois.ripe.net',
  '92.0.0.0/8' => 'whois.ripe.net',
  '93.0.0.0/8' => 'whois.ripe.net',
  '94.0.0.0/8' => 'whois.ripe.net',
  '95.0.0.0/8' => 'whois.ripe.net',
  '96.0.0.0/8' => 'whois.arin.net',
  '97.0.0.0/8' => 'whois.arin.net',
  '98.0.0.0/8' => 'whois.arin.net',
  '99.0.0.0/8' => 'whois.arin.net',
  '100.0.0.0/8' => 'whois.arin.net',
  '101.0.0.0/8' => 'whois.apnic.net',
  '102.0.0.0/8' => 'whois.afrinic.net',
  '103.0.0.0/8' => 'whois.apnic.net',
  '104.0.0.0/8' => 'whois.arin.net',
  '105.0.0.0/8' => 'whois.afrinic.net',
  '106.0.0.0/8' => 'whois.apnic.net',
  '107.0.0.0/8' => 'whois.arin.net',
  '108.0.0.0/8' => 'whois.arin.net',
  '109.0.0.0/8' => 'whois.ripe.net',
  '110.0.0.0/8' => 'whois.apnic.net',
  '111.0.0.0/8' => 'whois.apnic.net',
  '112.0.0.0/8' => 'whois.apnic.net',
  '113.0.0.0/8' => 'whois.apnic.net',
  '114.0.0.0/8' => 'whois.apnic.net',
  '115.0.0.0/8' => 'whois.apnic.net',
  '116.0.0.0/8' => 'whois.apnic.net',
  '117.0.0.0/8' => 'whois.apnic.net',
  '118.0.0.0/8' => 'whois.apnic.net',
  '119.0.0.0/8' => 'whois.apnic.net',
  '120.0.0.0/8' => 'whois.apnic.net',
  '121.0.0.0/8' => 'whois.apnic.net',
  '122.0.0.0/8' => 'whois.apnic.net',
  '123.0.0.0/8' => 'whois.apnic.net',
  '124.0.0.0/8' => 'whois.apnic.net',
  '125.0.0.0/8' => 'whois.apnic.net',
  '126.0.0.0/8' => 'whois.apnic.net',
  '128.0.0.0/8' => 'whois.arin.net',
  '129.0.0.0/8' => 'whois.arin.net',
  '130.0.0.0/8' => 'whois.arin.net',
  '131.0.0.0/8' => 'whois.arin.net',
  '132.0.0.0/8' => 'whois.arin.net',
  '133.0.0.0/8' => 'whois.apnic.net',
  '134.0.0.0/8' => 'whois.arin.net',
  '135.0.0.0/8' => 'whois.arin.net',
  '136.0.0.0/8' => 'whois.arin.net',
  '137.0.0.0/8' => 'whois.arin.net',
  '138.0.0.0/8' => 'whois.arin.net',
  '139.0.0.0/8' => 'whois.arin.net',
  '140.0.0.0/8' => 'whois.arin.net',
  '141.0.0.0/8' => 'whois.ripe.net',
  '142.0.0.0/8' => 'whois.arin.net',
  '143.0.0.0/8' => 'whois.arin.net',
  '144.0.0.0/8' => 'whois.arin.net',
  '145.0.0.0/8' => 'whois.ripe.net',
  '146.0.0.0/8' => 'whois.arin.net',
  '147.0.0.0/8' => 'whois.arin.net',
  '148.0.0.0/8' => 'whois.arin.net',
  '149.0.0.0/8' => 'whois.arin.net',
  '150.0.0.0/8' => 'whois.apnic.net',
  '151.0.0.0/8' => 'whois.ripe.net',
  '152.0.0.0/8' => 'whois.arin.net',
  '153.0.0.0/8' => 'whois.apnic.net',
  '154.0.0.0/8' => 'whois.afrinic.net',
  '155.0.0.0/8' => 'whois.arin.net',
  '156.0.0.0/8' => 'whois.arin.net',
  '157.0.0.0/8' => 'whois.arin.net',
  '158.0.0.0/8' => 'whois.arin.net',
  '159.0.0.0/8' => 'whois.arin.net',
  '160.0.0.0/8' => 'whois.arin.net',
  '161.0.0.0/8' => 'whois.arin.net',
  '162.0.0.0/8' => 'whois.arin.net',
  '163.0.0.0/8' => 'whois.apnic.net',
  '164.0.0.0/8' => 'whois.arin.net',
  '165.0.0.0/8' => 'whois.arin.net',
  '166.0.0.0/8' => 'whois.arin.net',
  '167.0.0.0/8' => 'whois.arin.net',
  '168.0.0.0/8' => 'whois.arin.net',
  '169.0.0.0/8' => 'whois.arin.net',
  '170.0.0.0/8' => 'whois.arin.net',
  '171.0.0.0/8' => 'whois.apnic.net',
  '172.0.0.0/8' => 'whois.arin.net',
  '173.0.0.0/8' => 'whois.arin.net',
  '174.0.0.0/8' => 'whois.arin.net',
  '175.0.0.0/8' => 'whois.apnic.net',
  '176.0.0.0/8' => 'whois.ripe.net',
  '177.0.0.0/8' => 'whois.lacnic.net',
  '178.0.0.0/8' => 'whois.ripe.net',
  '179.0.0.0/8' => 'whois.lacnic.net',
  '180.0.0.0/8' => 'whois.apnic.net',
  '181.0.0.0/8' => 'whois.lacnic.net',
  '182.0.0.0/8' => 'whois.apnic.net',
  '183.0.0.0/8' => 'whois.apnic.net',
  '184.0.0.0/8' => 'whois.arin.net',
  '185.0.0.0/8' => 'whois.ripe.net',
  '186.0.0.0/8' => 'whois.lacnic.net',
  '187.0.0.0/8' => 'whois.lacnic.net',
  '188.0.0.0/8' => 'whois.ripe.net',
  '189.0.0.0/8' => 'whois.lacnic.net',
  '190.0.0.0/8' => 'whois.lacnic.net',
  '191.0.0.0/8' => 'whois.lacnic.net',
  '192.0.0.0/8' => 'whois.arin.net',
  '193.0.0.0/8' => 'whois.ripe.net',
  '194.0.0.0/8' => 'whois.ripe.net',
  '195.0.0.0/8' => 'whois.ripe.net',
  '196.0.0.0/8' => 'whois.afrinic.net',
  '197.0.0.0/8' => 'whois.afrinic.net',
  '198.0.0.0/8' => 'whois.arin.net',
  '199.0.0.0/8' => 'whois.arin.net',
  '200.0.0.0/8' => 'whois.lacnic.net',
  '201.0.0.0/8' => 'whois.lacnic.net',
  '202.0.0.0/8' => 'whois.apnic.net',
  '203.0.0.0/8' => 'whois.apnic.net',
  '204.0.0.0/8' => 'whois.arin.net',
  '205.0.0.0/8' => 'whois.arin.net',
  '206.0.0.0/8' => 'whois.arin.net',
  '207.0.0.0/8' => 'whois.arin.net',
  '208.0.0.0/8' => 'whois.arin.net',
  '209.0.0.0/8' => 'whois.arin.net',
  '210.0.0.0/8' => 'whois.apnic.net',
  '211.0.0.0/8' => 'whois.apnic.net',
  '212.0.0.0/8' => 'whois.ripe.net',
  '213.0.0.0/8' => 'whois.ripe.net',
  '216.0.0.0/8' => 'whois.arin.net',
  '217.0.0.0/8' => 'whois.ripe.net',
  '218.0.0.0/8' => 'whois.apnic.net',
  '219.0.0.0/8' => 'whois.apnic.net',
  '220.0.0.0/8' => 'whois.apnic.net',
  '221.0.0.0/8' => 'whois.apnic.net',
  '222.0.0.0/8' => 'whois.apnic.net',
  '223.0.0.0/8' => 'whois.apnic.net',
};
  return $self;
}

sub get_server {
  my $self = shift;
  my $ip = shift || return undef;

  foreach my $range ( keys %{ $self->{data} } ) {
	if ( $range eq '0.0.0.0/2' ) {
		foreach my $cls_a ( 1 .. 126 ) {
		  my $block2 = Net::Netmask->new( "$cls_a.0.0.0/8" );
		  if ( $block2->match( $ip ) ) {
			return ( $self->{data}->{ $range }, $range );
		  }
		}
	}
	my $block = Net::Netmask->new( $range );
	if ( $block->match( $ip ) ) {
		return ( $self->{data}->{ $range }, $range );
	}
  }
  return undef;
}

1;

__END__

=pod

=head1 NAME

POE::Component::Client::Whois::IPBlks - Determine which whois server is responsible for a network address.

=head1 VERSION

version 1.30

=head1 SYNOPSIS

  use strict;
  use POE::Component::Client::Whois::IPBlks;

  my $ipblks = POE::Component::Client::Whois::IPBlks->new();

  my $whois_server = $ipblks->get_server('192.168.1.12');

=head1 DESCRIPTION

POE::Component::Client::Whois::IPBlks provides the ability to determine which whois server is responsible for a network address. It has a list of network ranges mapped to whois servers and uses L<Net::Netmask> to determine the appropriate Whois server for the given address.

=head1 CONSTRUCTOR

=over

=item C<new>

Returns a POE::Component::Client::Whois::IPBlks object.

=back

=head1 METHODS

=over

=item C<get_server>

Takes a single argument, an IP address to lookup the Whois for. Returns the applicable whois server or undef on failure.

=back

=head1 SEE ALSO

L<Net::Netmask>

L<http://www.iana.org/assignments/ipv4-address-space/ipv4-address-space.xhtml>

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Chris Williams.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/POE/Component/Client/Whois.pm
package POE::Component::Client::Whois;
{
  $POE::Component::Client::Whois::VERSION = '1.30';
}

#ABSTRACT: A one shot non-blocking RFC 812 WHOIS query.

use strict;
use warnings;
use Socket;
use Carp;
use POE qw(Filter::Line Wheel::ReadWrite Wheel::SocketFactory);
use POE::Component::Client::Whois::TLDList;
use POE::Component::Client::Whois::IPBlks;

sub whois {
    my $package = shift;
    my %args    = @_;

    $args{ lc $_ } = delete $args{$_} for keys %args;

    $args{referral} = 1 unless defined $args{referral} and !$args{referral};

    unless ( $args{query} and $args{event} ) {
        warn "You must provide a query string and a response event\n";
        return undef;
    }

    unless ( $args{host} ) {
        my $whois_server;
        my $tld = POE::Component::Client::Whois::TLDList->new();
        my $blk = POE::Component::Client::Whois::IPBlks->new();
      SWITCH: {
            if ( $args{query} =~ /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/
                and
                scalar( grep $_ >= 0 && $_ <= 255, split /\./, $args{query} ) ==
                4 )
            {
                $whois_server = ( $blk->get_server( $args{query} ) )[0];
                unless ($whois_server) {
                    warn
"Couldn\'t determine correct whois server, falling back on arin\n";
                    $whois_server = 'whois.arin.net';
                }
                last SWITCH;
            }
            if ( $args{query} =~ /:/ ) {
                warn "IPv6 detected, defaulting to 6bone\n";
                $whois_server = 'whois.6bone.net';
                last SWITCH;
            }
            $whois_server = ( $tld->tld( $args{query} ) )[0];
            if ( $whois_server eq 'ARPA' ) {
                $args{query} =~ s/\.in-addr\.arpa//;
                $args{query} = join '.', reverse split( /\./, $args{query} );
                $whois_server = ( $blk->get_server( $args{query} ) )[0];
                unless ($whois_server) {
                    warn
"Couldn\'t determine correct whois server, falling back on arin\n";
                    $whois_server = 'whois.arin.net';
                }
            }
            unless ($whois_server) {
                warn
"Could not automagically determine whois server from query string, defaulting to internic \n";
                $whois_server = 'whois.internic.net';
            }
        }
        $args{host} = $whois_server;
    }

    $args{session} = $poe_kernel->get_active_session()
      unless ( $args{session} );

    my $self = bless { request => \%args }, $package;

    $self->{session_id} = POE::Session->create(
        object_states => [
            $self => [
                qw(_start _connect _sock_input _sock_down _sock_up _sock_failed _time_out)
            ],
        ],
        options => { trace => 0 },
    )->ID();

    return $self;
}

sub _start {
    my ( $kernel, $self ) = @_[ KERNEL, OBJECT ];
    $self->{_dot_com} = ( POE::Component::Client::Whois::TLDList->new()->tld('.com') )[0];
    $self->{session_id} = $_[SESSION]->ID();
    $kernel->yield('_connect');
    undef;
}

sub _connect {
    my ( $kernel, $self ) = @_[ KERNEL, OBJECT ];

    # Check here for NONE or WEB and send an error straight away.
    if ( my ($type) = $self->{request}->{host} =~ /^(NONE|WEB)$/ ) {
        my $error;
        if ( $type eq 'NONE' ) {
            $error = 'This TLD has no whois server.';
        }
        else {
            $error =
                'This TLD has no whois server, but you can access the '
              . 'whois database at '
              . (
                POE::Component::Client::Whois::TLDList->new->tld(
                    $self->{request}->{query}
                )
              )[1];
        }
        $self->{request}->{error} = $error;
        my $request = delete $self->{request};
        my $session = delete $request->{session};
        $kernel->post( $session => $request->{event} => $request );
        return;
    }
    $self->{factory} = POE::Wheel::SocketFactory->new(
        SocketDomain   => AF_INET,
        SocketType     => SOCK_STREAM,
        SocketProtocol => 'tcp',
        RemoteAddress  => $self->{request}->{host},
        RemotePort     => $self->{request}->{port} || 43,
        SuccessEvent   => '_sock_up',
        FailureEvent   => '_sock_failed',
    );
    undef;
}

sub _sock_failed {
    my ( $kernel, $self, $op, $errno, $errstr ) =
      @_[ KERNEL, OBJECT, ARG0 .. ARG2 ];

    delete $self->{factory};
    $self->{request}->{error} = "$op error $errno: $errstr";
    my $request = delete $self->{request};
    my $session = delete $request->{session};

    $kernel->post( $session => $request->{event} => $request );
    undef;
}

sub _sock_up {
    my ( $kernel, $self, $session, $socket ) =
      @_[ KERNEL, OBJECT, SESSION, ARG0 ];
    delete $self->{factory};

    $self->{'socket'} = new POE::Wheel::ReadWrite(
        Handle => $socket,
        Driver => POE::Driver::SysRW->new(),
        Filter => POE::Filter::Line->new(
            InputRegexp   => '\015?\012',
            OutputLiteral => "\015\012"
        ),
        InputEvent => '_sock_input',
        ErrorEvent => '_sock_down',
    );

    unless ( $self->{'socket'} ) {
        my $request = delete $self->{request};
        my $session = delete $request->{session};
        $request->{error} =
          "Couldn\'t create a Wheel::ReadWrite on socket for whois";
        $kernel->post( $session => $request->{event} => $request );
        return undef;
    }

    $self->{'socket'}->put( $self->{request}->{query} );
    $kernel->delay( '_time_out' => 30 );
    undef;
}

sub _sock_down {
    my ( $kernel, $self ) = @_[ KERNEL, OBJECT ];
    delete $self->{socket};
    $kernel->delay( '_time_out' => undef );

    if ( $self->{request}->{referral} and $self->{_referral} ) {
        delete $self->{request}->{reply} if $self->{referral_only};
        $self->{request}->{host} = delete $self->{_referral};
        $kernel->yield('_connect');
        return;
    }

    my $request = delete $self->{request};
    my $session = delete $request->{session};

    if ( defined( $request->{reply} ) and ref( $request->{reply} ) eq 'ARRAY' )
    {
        delete $request->{error};
    }
    else {
        $request->{error} = "No information received from remote host";
    }
    $kernel->post( $session => $request->{event} => $request );
    undef;
}

sub _sock_input {
    my ( $kernel, $self, $line ) = @_[ KERNEL, OBJECT, ARG0 ];
    push @{ $self->{request}->{reply} }, $line;
    if ( my ($referral) = $line =~ /ReferralServer:\s+(.*)$/ ) {
        my ( $scheme, $authority, $path, $query, $fragment ) = $referral =~
          m|(?:([^:/?#]+):)?(?://([^/?#]*))?([^?#]*)(?:\?([^#]*))?(?:#(.*))?|;
        return unless $scheme and $authority;
        $scheme = lc $scheme;
        return unless $scheme eq 'whois';
        my ( $host, $port ) = split /:/, $authority;
        return if $host eq $self->{request}->{host};
        $self->{_referral} = $host;
    }
    if ( $self->{request}->{host} eq $self->{_dot_com} 
        and my ($other) = $line =~ /Whois Server:\s+(.*)\s*$/i )
    {
        $self->{_referral} = $other;
    }
    undef;
}

sub _time_out {
    my ( $kernel, $self ) = @_[ KERNEL, OBJECT ];
    delete $self->{'socket'};
    undef;
}

1;

__END__

=pod

=head1 NAME

POE::Component::Client::Whois - A one shot non-blocking RFC 812 WHOIS query.

=head1 VERSION

version 1.30

=head1 SYNOPSIS

   use strict;
   use warnings;
   use POE qw(Component::Client::Whois);
   use Data::Dumper;

   POE::Session->create(
	package_states => [
		'main' => [ qw(_start _response) ],
	],
   );

   $poe_kernel->run();
   exit 0;

   sub _start {
     my ($kernel,$heap) = @_[KERNEL,HEAP];

     POE::Component::Client::Whois->whois( host => "whois.nic.uk",
					   query => 'bingosnet.co.uk',
					   event => '_response',
					   _arbitary => [ qw(moo moo moo) ] );
     undef;
   }

   sub _response {
      print STDERR Dumper( $_[ARG0] );
   }

=head1 DESCRIPTION

POE::Component::Client::Whois provides a lightweight one shot non-blocking RFC 812 WHOIS query to other POE sessions and components. The component will attempt to guess the appropriate whois server to connect to based on the query string passed.

If no guess can be made it will connect to whois.internic.net for
domains, whois.arin.net for IPv4 addresses and whois.6bone.net for IPv6  addresses.

=head1 CONSTRUCTOR

=over

=item C<whois>

Creates a POE::Component::Client::Whois session. Takes two mandatory arguments and a number of optional:

  'query', the string query to send to the whois server; # Mandatory
  'event', the event name to emit on success/failure; # Mandatory
  'port', the port on the whois server to connect to, defaults to 43;
  'session', a session or alias to send the above 'event' to, defaults to calling session;
  'host', the whois server to query; # Automagically determined by the component
  'referral', indicates to the poco whether to follow ReferralServer; # Default is 1
  'referral_only', indicates whether the poco should only return the results of querying
		   the referral server; # default is 0

One can also pass arbitary data to whois() which will be passed back in the response event. It is advised that one uses
an underscore prefix to avoid clashes with future versions.

=back

=head1 OUTPUT

ARG0 will be a hashref, which contains the original parameters passed to whois() ( including any arbitary data ), plus either one of the following two keys:

  'reply', an arrayref of response lines from the whois server, assuming no error occurred;
  'error', in lieu of a valid response, this will be defined with a brief description of
	   what went wrong;

No parsing is undertaken on the returned data, this is an exercise left to the reader >;]

=head1 KUDOS

ketas, for first suggesting this module;
buu, decay and hazard from #perl @ freenode, for helpful suggestions;

This module is based on the linux whois client from L<http://www.linux.it/~md/software/>.

=head1 SEE ALSO

RFC 812 L<http://www.faqs.org/rfcs/rfc812.html>.

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Chris Williams.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
